### PR TITLE
Timeseries: Add hover proximity option

### LINKED ISSUE #6946: Show only active series under the cursor in tooltip
Would like a new option to enable displaying of only active series under cursor when hovering a chart.

Hover info: [All series, Single, Present] or something like that.

screen shot 2016-12-13 at 18 33 44

### PR DESCRIPTION
**Pull request by adela-almasan**
Before

https://github.com/grafana/grafana/assets/88068998/0e0d3b5d-8b03-489b-9c95-cf4360485a11

After

https://github.com/grafana/grafana/assets/88068998/8cb43b8b-999a-4e2e-8f6d-730be3ffae7d

Fixes https://github.com/grafana/support-escalations/issues/9081
Fixes https://github.com/grafana/grafana/issues/6946

TODO
- [x] Add `hoverProximity` option to Timeseries, Candlestick and Trend
- [x] Filter null values from tooltip
- [x] Trend flicker?
- [x] Update uPlot version
- [x] Update docs

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by leeoniya** [comment]
i wonder if it's worth being explicit that this controls "x" proximity only?

we have a different, internal `cursor.focus.prox` setting which determines when to trigger the tooltip and series focus in the "y" direction.

File: docs/sources/panels-visualizations/visualizations/time-series/index.md [hunk 1]
```diff
@@ -50,6 +50,10 @@ Tooltip options control the information overlay that appears when you hover over
 
 {{< docs/shared lookup="visualizations/tooltip-mode.md" source="grafana" version="<GRAFANA VERSION>" >}}
 
+### Hover proximity
+
+This option controls how close your cursor must be to a data point before the tooltip appears.
```
Comment (imatwawana) [inline docs/sources/panels-visualizations/visualizations/time-series/index.md:55]: I feel like we should be explicit the measurement about this as I've definitely wondered about this at times. Feel free to change my wording.
```suggestion
This option controls how close your cursor must be to a data point before the tooltip appears. Values are in pixels.
```

**Review by imatwawana — APPROVED** [review]
One nit, but otherwise looks good to me.

File: packages/grafana-ui/src/options/builder/tooltip.tsx [hunk 2]
```diff
@@ -43,7 +44,21 @@ export function addTooltipOptions<T extends OptionsWithTooltip>(
       settings: {
         options: sortOptions,
       },
-    })
+    });
+
+  if (setProximity) {
+    builder.addNumberInput({
```
Comment (nmarrs) [inline packages/grafana-ui/src/options/builder/tooltip.tsx:50]: should we default this to `15` as per https://github.com/grafana/grafana/pull/81421/files#diff-8d24e749f24ed73b06db210b373dae9a91e859ea27635d0842e4fb308d288529R562?

**Review by nmarrs — COMMENTED** [review]
I agree with Leon that we probably should specify that this proximity option only affects the x axis. That being said is there a reason why we couldn't just apply this proximity to work in both x + y directions? That seems like the option would behave closer to what I would expect in this case 

https://github.com/grafana/grafana/assets/22381771/61a2e32a-ec14-4e12-a5f0-cfa977dcba53

**Review by baldm0mma — APPROVED** [review]

**Review by codeincarnate — APPROVED** [review]
Looks good overall! One thing to consider is we may want to make highlighted datapoints more obvious as it's potential with this that the tooltip can be rather far away from the highlighted datapoint. Doesn't need to be a part of this PR necessarily, though that would be a nice bonus 😄

File: packages/grafana-ui/src/components/uPlot/config/UPlotConfigBuilder.ts [hunk 3]
```diff
@@ -313,6 +313,7 @@ type UPlotConfigPrepOpts<T extends Record<string, unknown> = {}> = {
   tweakAxis?: (opts: AxisProps, forField: Field) => AxisProps;
   // Identifies the shared key for uPlot cursor sync
   eventsScope?: string;
+  hoverProximity?: number | undefined;
```
Comment (leeoniya) [inline packages/grafana-ui/src/components/uPlot/config/UPlotConfigBuilder.ts:316]: `?` implies `undefined`, so you don't need `| undefined` here

File: packages/grafana-ui/src/options/builder/tooltip.tsx [hunk 4]
```diff
@@ -4,6 +4,7 @@ import { OptionsWithTooltip, TooltipDisplayMode, SortOrder } from '@grafana/sche
 export function addTooltipOptions<T extends OptionsWithTooltip>(
   builder: PanelOptionsEditorBuilder<T>,
   singleOnly = false,
+  setProximity = false,
```
Comment (leeoniya) [inline packages/grafana-ui/src/options/builder/tooltip.tsx:7]: this is called a [boolean trap](https://ariya.io/2011/08/hall-of-api-shame-boolean-trap) :)

you end up with `addTooltipOptions(builder, false, true, opts);`, which is a confusing API

not worth cleaning it up in this PR, but might consider switching to a more explicit props object in the future.

**Review by leeoniya — APPROVED** [review]
small nit about unneeded `| undefined`, but LGTM :+1:

**Pull request merged**